File: 303 Range Sum Query - Immutable/range_sum_query_immutable.cpp
class NumArray {
public:
    NumArray(vector<int> &nums) {
        int sum = 0;
        sums.push_back(0);
        for (auto n : nums) {
            sum += n;
            sums.push_back(sum);
        }
    }

    int sumRange(int i, int j) {
        return sums[j + 1] - sums[i];
    }
private:
    vector<int> sums;
};


// Your NumArray object will be instantiated and called as such:
// NumArray numArray(nums);
// numArray.sumRange(0, 1);
// numArray.sumRange(1, 2);

File: 303 Range Sum Query - Immutable/range_sum_query_immutable.cs
public class NumArray	// Author: Xinyu Chen
{
    private int[] data;
    
    public NumArray(int[] nums)
    {
        if (nums.Length == 0) return;
        data = new int[nums.Length];
        data[0] = nums[0];
        for (int i = 1; i < nums.Length; i++)
        {
            data[i] = data[i - 1] + nums[i];
        }
    }
    
    public int SumRange(int i, int j)
    {
        if (i == 0) return data[j];
        return data[j] - data[i - 1];
    }
}


// Your NumArray object will be instantiated and called as such:
// NumArray numArray = new NumArray(nums);
// numArray.SumRange(0, 1);
// numArray.SumRange(1, 2);
